feat: add allowTypeAnnotation to func-style (#19754)
eslint/eslint: docs/src/rules/func-style.md

This rule has an object option for two exceptions:

* `"allowArrowFunctions"`: `true` (default `false`) allows the use of arrow functions. This option applies only when the string option is set to `"declaration"` (arrow functions are always allowed when the string option is set to `"expression"`, regardless of this option)

> * `"allowTypeAnnotation"`: `true` (default `false`) allows the use of function expressions and arrow functions when they have type annotations, regardless of the `allowArrowFunctions` option. This option applies only when the string option is set to `"declaration"`. (TypeScript only)

Reviewer: ```suggestion
* `"allowTypeAnnotation"`: `true` (default `false`) allows the use of function expressions and arrow functions when the variable declaration has type annotation, regardless of the `allowArrowFunctions` option. This option applies only when the string option is set to `"declaration"`. (TypeScript only)
```

I think this would be more accurate.

:::

### allowTypeAnnotation

Examples of additional **correct** TypeScript code for this rule with the `"declaration", { "allowTypeAnnotation": true }` options:

::: correct

```ts
/*eslint func-style: ["error", "declaration", { "allowTypeAnnotation": true }]*/

> const foo: () => void = function() {};
> 
> const bar: () => void = () => {};

Reviewer: ```suggestion
const foo: Fn = function() {};

const bar: Fn = () => {};
```

I think examples where an existing type is used would be more representative for the purpose of this option.
